src/main/scala/is/hail/asm4s/CodeWithRefs.scala

```diff
@@ -0,0 +1,26 @@
+package is.hail.asm4s
+
+import scala.language.implicitConversions
+
+case class CodeWithRefs[T](action: FunctionBuilder[_] => T) {
+  def map[U](f: T => U): CodeWithRefs[U] =
+    CodeWithRefs(fb => f(action(fb)))
+  def flatMap[U](f: T => CodeWithRefs[U]): CodeWithRefs[U] =
+    CodeWithRefs(fb => f(action(fb)).action(fb))
+  def withFilter(p: T => Boolean): CodeWithRefs[T] =
+    CodeWithRefs(fb => { val t = action(fb); if (p(t)) t else throw new RuntimeException() })
+  def filter(p: T => Boolean): CodeWithRefs[T] =
+    withFilter(p)
+  def run[R >: Null](fb: Function0Builder[R])(implicit ev: T <:< Code[R]): () => R =
```

Comment:
We need to generalize the result/run interface so we don't have to specialize it for each type.  That can go in a separate PR if you like.